src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -1986,12 +1987,24 @@ private function registerMailerConfiguration(array $config, ContainerBuilder $co
             }
         }
 
-        $recipients = $config['envelope']['recipients'] ?? null;
-        $sender = $config['envelope']['sender'] ?? null;
-
         $envelopeListener = $container->getDefinition('mailer.envelope_listener');
-        $envelopeListener->setArgument(0, $sender);
-        $envelopeListener->setArgument(1, $recipients);
+        $envelopeListener->setArgument(0, $config['envelope']['sender'] ?? null);
+        $envelopeListener->setArgument(1, $config['envelope']['recipients'] ?? null);
+
+        if ($config['headers']) {
+            $headers = new Headers();
+            foreach ($config['headers'] as $name => $data) {
+                $value = $data['value'];
+                if (\in_array($name, ['from', 'to', 'cc', 'bcc', 'reply-to'])) {
+                    $value = (array) $value;
+                }
+                $headers->addHeader($name, $value);
+            }
+            $messageListener = $container->getDefinition('mailer.message_listener');
+            $messageListener->setArgument(0, $headers);
```

Comment:
Does this work, passing already created instances of `Headers`? Should these be defined using definitions instead?

Reply:
fixed :)